Distinguish type arguments vs less/greater, fixes #81

Ptival opened the pull request:
As explained in #81, the current indentation fails to discern type arguments like `t < a , b > ...` from comparison operators separated by commas like `t < a , b > ...`.

It is a somewhat hard problem to solve using only regular-expressions, yet, this patch solves all of the tricky examples listed in #81, many of which are, honestly, code that should not be written without parentheses!

So this patch:
- keeps all existing good behaviors
- patches most (all?) bad behaviors for reasonable code
- patches many(?) bad behaviors for unreasonable code

And I've kept it short and documented, so that it should not become a maintenance burden.

Cheers!

Ptival commented:
Oh, I did not notice that file. I will add them. Is the intent that, the file, as it is, is correctly indented, and so running the indentation script over it should be a no-op?

lddubeau commented:
Yes, running the indentation script over the reference file should be a no-op.

Ptival commented:
Ready for review!

lddubeau commented:
`typescript--symbols-forbidden-in-type-parameters-re` is incorrect. I quickly came up with cases where `+`, `-` and `=` appear in type parameters. The following compiles without error with the latest `tsc`. The indentation is thrown off by the appearance of `+`, `-` and `=` in type parameters (see the lines with `const moo` and `const moo2`).

I've not found other cases (with `/`, `*`, `%`, etc.) but that does not mean they don't exist.

```
// Good case

class Q<X> {
    q: string = "a"
}

type a<X> =
    Q<X>
const blah = 1;

// Problem cases

interface Something {
    a: string;
    
    b: string;

    c: -5;
}

class Fluff<X extends Something> {
}

// Example of = and - in a type parameter.
type c<X extends Something = { a: string; b: string; c: -5; more: string }>
    = Fluff<X>
    
    const moo = 1

// Example of + in a type parameter.
type d<X extends Something = { +readonly [P in keyof Something]: Something[P] }>
    = Fluff<X>
    
    const moo2 = 1
```

Ptival commented:
My bad indeed! I have removed this incorrect "optimization", and now the code passes all my tests, and I believe, all the ones you added (I also added them to the test file).

Ptival commented:
Unfortunately, my previous incorrect optimization made it so that, when we are looking for a matching angle bracket, but there's none to be found, we end up going way deep into recursion while searching backwards.

I'd like other people's input on this, but I believe that if we see one of those keywords, we are definitely crossing by a sentence starter, thus, we should stop looking for the angle bracket:
```
const
export
function
import
let
type
var
```

I have tried to convince myself of this, indeed those words cannot, at the moment, be used as variable names. But, it could be the case that, through some syntax I am not aware of, or not thinking about, those could appear in a type annotation. Let me know if you think of a counter-example.

If you believe this is correct, my last commit adds this as a way of bailing out of the search for the opening matching bracket, whenever our search encounters one of those keywords first.

All the examples in the test-file for indentation still work, and it also solved my problem in a very large file where the recursion depth would get too deep.

Ptival commented:
Hold off this PR a little, I have just encountered a case I had not thought of, will try and fix it shortly.

lddubeau commented:
@Ptival I saw you made an additional commit. Does this mean that you've taken care of the problem you saw, and it is ready for review?

Ptival commented:
Yes, that last commit should have fixed that. I also removed some `(backward-char)` calls that were erroneous, now that I understand better how the point is moving via the different actions. Ready for review!

lddubeau commented on typescript-mode.el:
```diff
@@ -2056,6 +2089,45 @@ This performs fontification according to `typescript--class-styles'."
+                  (typescript--re-search-backward (concat "[=:]\\|" typescript--keyword-re) nil t)
+                  (or
+                   ;; If the previous keyword is "as", definitely a type.
+                   (when (looking-at (typescript--regexp-opt-symbol '("as"))) t)
```
Prior to your commit, all calls to `typescript--regexp-opt-symbol` are made in the initialization of global variables. And elsewhere the regexps are just constructed manually. I'd rather keep it that way and avoid the call in code that performs performance sensitive tasks. From time to time I have to use modes that are slow at indentation. I definitely notice when Emacs has to pause to indent my line.

Here you can easily replace `(typescript--regexp-opt-symbol '("as"))` with `\\_<as\\_>`

lddubeau commented on typescript-mode.el:
```diff
@@ -2056,6 +2089,45 @@ This performs fontification according to `typescript--class-styles'."
+                   (when (looking-at ":")
+                     (and
+                      (typescript--re-search-backward
+                       (concat "[?]\\|" (typescript--regexp-opt-symbol '("as" "class" "private" "public" "readonly" "type")))
```
The entire `concat` can be replaced with a precomputed string. 

Also, `"type"` in this list might face the same problem as the issue I raised on `typescript--reserved-start-keywords-re`. If it was included here for the same reason it was included there, then it should be removed here too.

lddubeau commented on typescript-mode.el:
```diff
@@ -2056,6 +2089,45 @@ This performs fontification according to `typescript--class-styles'."
+                   (when (looking-at "=")
+                     (and
+                      (typescript--re-search-backward-ignoring-angle-brackets (concat "[=}]\\|" (typescript--regexp-opt-symbol '("type"))))
+                      (when (looking-at (typescript--regexp-opt-symbol '("type"))) t))))))))
```
The two previous lines contain trivial calls to `typescript--regexp-opt-symbol` which ought to be replace with precomputed strings.

Also, this is another case of `(when test t)` which can be reduced to `test`.

lddubeau commented on typescript-mode.el:
```diff
@@ -2031,6 +2031,39 @@ This performs fontification according to `typescript--class-styles'."
 
+(defconst typescript--reserved-start-keywords-re
+    (typescript--regexp-opt-symbol '("const" "export" "function" "import" "let" "type" "var"))
+    "These keywords cannot be variable or type names and start a new sentence.")
```
`type` can definitely be a variable or a type name. I can compile this without any issue. Here's an [example](https://www.typescriptlang.org/play/#src=type%20type%20%3D%20number%3B%0D%0A%0D%0Aconst%20type%3A%20type%20%3D%201%3B%0D%0A) on the playground. It does not raise any compilation errors.

I tried just removing `type` and your code still worked. I suppose we could go ahead without `type` in this list. As I understand it, these are just meant to allow stopping searches early.

lddubeau commented:
@Ptival I see a commit. Thanks for the update. However, I see that there are 3 changes requested in my previous review that are listed as "outdated" rather than "resolved". Did you set them aside and plan to deal with them later?

Ptival commented:
@lddubeau Ha! No, I missed them, they were folded and I unintentionally skipped over them. Will try to address those soon!

Ptival commented on typescript-mode.el:
```diff
@@ -2031,6 +2031,39 @@ This performs fontification according to `typescript--class-styles'."
+    (typescript--regexp-opt-symbol '("const" "export" "function" "import" "let" "type" "var"))
+    "These keywords cannot be variable or type names and start a new sentence.")
+
+(defun typescript--search-backward-matching-angle-bracket (depth)
```
I definitely agree with this.

Does it make sense to make the inner function be locally-scoped within the exposed one? Can you achieve this in elisp by just doing something like:
```
(defun outer
  (defun inner (depth) ...)
  (inner 1))
```
?

Or is there a preferred way or doing this?

Ptival replied:
Oh, would this be a bad idea in elisp, because each call would re-define the function? I'll just follow the examples from the rest of the file and define the `inner` separately.

lddubeau replied:
You ended up going the way I think is best but here are a couple of observations on this question.

I would not do `defun` inside `defun` for this. `defun` is really meant for defining "top-level" functions rather than defining local functions.  You'd use something like `cl-flet` to create a local binding. (There's also `flet` but `flet` is obsolete.)

And yes each call would create the function anew, which may have performance implications.

Given that Emacs operates largely through a "private by convention" model rather than enforce privacy trough language features, I'd do it the way you settled on: with `...-inner`.

Ptival commented:
Ok, i committed changes for those 3 remaining change requests!

Ptival commented:
I have added two edge cases that I came upon when re-indenting my whole code base. They don't add much complexity (to either the code or the execution), but have a look!

josteink commented:
This PR just never ends, does it? 😄 

Good job though @Ptival, helping out making typescript-mode a better place for everyone 👍 

There's a slight conflict against the test-document in master (after your each patch). Could you resolve that as well, so that the PR is mergable?

lddubeau commented on typescript-mode.el:
```diff
@@ -2031,6 +2031,53 @@ This performs fontification according to `typescript--class-styles'."
+       (case (char-after)
+         (?< (typescript--search-backward-matching-angle-bracket-inner (- depth 1)))
+         (?> (typescript--search-backward-matching-angle-bracket-inner (+ depth 1)))
+         (otherwise nil)))))
```
I'm pretty sure the explicit `(otherwise nil)` is redundant. If a `case` does not match any of its branches, then `nil` is returned by default.

Ptival commented:
Done! :-)

lddubeau commented:
Thanks. It is now merged!
